core/src/sumeragi/main_loop.rs

```diff
@@ -1140,12 +1184,22 @@ pub(crate) fn run(
 
             // NOTE: View change must be periodically suggested until it is accepted.
             // Must be initialized to pipeline time but can increase by chosen amount
-            view_change_time += state.world.view().parameters().sumeragi.pipeline_time();
+            view_change_time += state
+                .world
+                .view()
+                .parameters()
+                .sumeragi
+                .pipeline_time(view_change_index, sumeragi.topology.max_faults() + 1);
```

Comment:
```suggestion
                .pipeline_time(view_change_index, sumeragi.topology.max_faults() + 1);
```

IMO `max_faults` is too much. Let's hardcode to some smaller number for now

Reply:
It would be 2 for 4 network and 3 for 7 node network.
What in your opinion would be reasonable number?

Reply:
and 33 for 100. That's pretty big. I'd suggest maybe 5 as the upper limit? We could try to make some calculation. What would have to be the value of `shift` so that probability that failure is due to a malicious node is less than p? or something along those lines

Reply:
If we assume that every faulty node try to submit genesis peer than probability that given peer would submit incorrect block is ~1/3.
And shift can't increase or decrease this probability only change time when block will be committed.

If we indeed have large block than approach with shift is actually worse by factor `shift` * `commit_time`.

In case where we have `f` wrong blocks in a row and than valid approach with a shift behave better by:
`diff = commit_time * (f + 2) * (f +1) / 2 - commit_time * (f + 1) = commit_time * (f + 1) * (f - 1) / 2`.

So the difference is quadratic of `f`.

Examples (commit_time=1s):
- f=1 => diff=0s
- f=2 => diff=1.5s
- f=3 => diff=4s
- f=4 => diff=7.5s
- f=33 => diff=544s

Reply:
So having fixed shift might theoretically slightly improve latency for the case where we have multiple invalid blocks in a row.

Reply:
When faulty peers (including faulty leader i.e. malicious block) are least concerned, the shift would be 0. When most concerned, the shift would be 2f+1 (sufficient rotations to exclude every faulty peer from the validating set).

After all, since no theoretical weights can be assigned between the causes (complexity and disagreement), the only way would be to determine a constant

Reply:
@SamHSmith

Reply:
@Erigara @mversic My suggestion is that we add a parameter that represents the target maximum amount of faulty nodes. Then the equation can use that. `total_time = block_time + commit_time * max(1.0, view_change_index / max_faulty_count)`

The result is that pipeline time is constant when f < max_faulty_count. The time to commit the block is worst case linear with f. Until we surpass the target max fault count. Then pipeline time because non constant because something is clearly wrong. And the total time to commit the block becomes quadratic.

Reply:
Can you specify wdym by "represents the target maximum amount of faulty nodes"?

Reply:
I mean that the Iroha customer could be able to specify how many faulty nodes they want to tolerate in the normal case. We assume that all nodes are online all the time and something is very wrong if they are faulty. It could be that you expect 1-3 nodes to be down for maintenance at any given time in your 20 node network.

On the other hand, it might be that the correct course of action is having smart contracts that automatically remove faulty peers from the network. In that case we are targeting no faulty peers in an iroha network. We can still produce blocks at 1/3 number of peers being faulty but we don't expect to stay at that high amount of faultyness. Then the speed problem becomes less of an issue because we only need to produce a few blocks to start kicking the faulty nodes.

Reply:
And maybe for public chains that is combined with some kind of staking or something.

Reply:
I'm concerned that this might complicate things for users. 
And other than that we already have max allowed number of faulty peers (`f` out of `3f + 1`).

Reply:
we don't need this to be configurable